File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/default_arguments/java/UserFactory.java
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.default_arguments.java;

import pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.readability.java.User;

import java.util.List;

public class UserFactory {

    public User createUser(String firstName, String lastName) {
        // Java has no default arguments.
        // Overloaded methods are commonly used instead.
        return createUser(firstName, lastName, 18, List.of("USER"));
    }

    public User createUser(String firstName, String lastName, int age) {
        return createUser(firstName, lastName, age, List.of("USER"));
    }

    public User createUser(String firstName, String lastName, int age, List<String> roles) {
        return new User(firstName, lastName, age, roles);
    }
}

File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/basics/default_arguments/kotlin/UserFactory.kt
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.default_arguments.kotlin

import pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.basics.readability.kotlin.User

class UserFactory {

    fun createUser(
        firstName: String,
        lastName: String,
        age: Int = 18,
        roles: List<String> = listOf("USER")
    ): User {
        // Default values reduce the need for overloaded methods.
        return User(
            firstName = firstName,
            lastName = lastName,
            age = age,
            roles = roles
        )
    }
}